src/qiboml/operations/differentiation.py

```diff
@@ -1,17 +1,21 @@
+from typing import Optional, Union
+
 import numpy as np
-from qibo.backends import construct_backend
+import qibo
+import qibo.backends
 from qibo.config import raise_error
 from qibo.hamiltonians.abstract import AbstractHamiltonian
+from qibojit.backends import NumbaBackend
 
 
-def parameter_shift(
-    hamiltonian,
-    circuit,
-    parameter_index,
-    initial_state=None,
-    scale_factor=1,
-    nshots=None,
-    backend="qibojit",
+def one_parameter_shift(
+    hamiltonian: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    parameter_index: int,
+    initial_state: Optional[Union[np.ndarray, qibo.Circuit]] = None,
+    scale_factor: float = 1.0,
```

Comment:
Again partially true: scale factor takes into account the fact that one may want the derivative of `partial` of the angle.  E.g. an angle which is defined as $\theta' = \theta\cdot a + b$. Thus it cannot be tackled as a general scale factor, but as something which is related to the specific angle considered. 
Anyway, we should consider to use Mike's `Parameter` class here. That would be a nice improvement.

Reply:
This you can also do after the gradient being returned. And I'd avoid any further structure on top of that, especially the `Parameter` class.

Moreover, I'd really avoid addressing something like $\theta' = \theta\cdot a + b$ inside the derivation: this is fully against composition.
Even TensorFlow, Pytorch, JAX, and whatever other differentiation framework are composable: they have rules for computing the gradients of building blocks, and one for the composition, the *chain rule*. And this is the basis of all autodiff, because you can automatically compose derivatives.

Thus, obtain the gradient w.r.t. the gate parameters as they are defined. If the parameters are functions of other parameters, the gradients should be composed (layers could come before and even after a circuit).